File: backend/src/services/exportService.ts
import ExcelJS from "exceljs";
import { Document, Packer, Paragraph, HeadingLevel, Table, TableRow, TableCell, TextRun, WidthType } from "docx";
import { DEFAULT_ROLES, type ProjectDto } from "shared";

function effectiveHours(hours: number, bufferPercent: number): number {
  return hours * (1 + bufferPercent / 100);
}

function phaseBreakdownSheet(workbook: ExcelJS.Workbook, project: ProjectDto) {
  const sheet = workbook.addWorksheet("Phase Breakdown");
  sheet.columns = [
    { header: "Phase", key: "name", width: 26 },
    { header: "Hours", key: "hours", width: 10 },
    { header: "Buffer %", key: "buffer", width: 10 },
    { header: "Effective Hours", key: "effective", width: 16 },
    { header: "Start", key: "start", width: 14 },
    { header: "End", key: "end", width: 14 },
    { header: "Rationale", key: "rationale", width: 50 },
    { header: "Dependencies", key: "dependencies", width: 30 },
  ];
  sheet.getRow(1).font = { bold: true };
  for (const phase of project.phases) {
    sheet.addRow({
      name: phase.name,
      hours: phase.hours,
      buffer: phase.bufferPercent,
      effective: Number(effectiveHours(phase.hours, phase.bufferPercent).toFixed(1)),
      start: phase.startDate ? new Date(phase.startDate).toLocaleDateString() : "",
      end: phase.endDate ? new Date(phase.endDate).toLocaleDateString() : "",
      rationale: phase.rationale ?? "",
      dependencies: phase.dependencies ?? "",
    });
  }
  const totalRow = sheet.addRow({
    name: "Total",
    hours: project.phases.reduce((s, p) => s + p.hours, 0),
    effective: Number(project.phases.reduce((s, p) => s + effectiveHours(p.hours, p.bufferPercent), 0).toFixed(1)),
  });
  totalRow.font = { bold: true };
}

function roleHoursSheet(workbook: ExcelJS.Workbook, project: ProjectDto) {
  const sheet = workbook.addWorksheet("Role Hours");
  const roles = Array.from(
    new Set([...DEFAULT_ROLES, ...project.customRoles, ...project.phases.flatMap((p) => p.roleHours.map((rh) => rh.role))])
  );
  sheet.columns = [{ header: "Phase", key: "phase", width: 26 }, ...roles.map((r) => ({ header: r, key: r, width: 16 })), { header: "Total", key: "total", width: 12 }];
  sheet.getRow(1).font = { bold: true };

  const roleTotals: Record<string, number> = Object.fromEntries(roles.map((r) => [r, 0]));
  for (const phase of project.phases) {
    const row: Record<string, number | string> = { phase: phase.name };
    let phaseTotal = 0;
    for (const role of roles) {
      const hours = phase.roleHours.find((rh) => rh.role === role)?.hours ?? 0;
      row[role] = hours;
      roleTotals[role] += hours;
      phaseTotal += hours;
    }
    row.total = phaseTotal;
    sheet.addRow(row);
  }
  const totalRow: Record<string, number | string> = { phase: "Total", total: Object.values(roleTotals).reduce((s, v) => s + v, 0) };
  for (const role of roles) totalRow[role] = roleTotals[role];
  sheet.addRow(totalRow).font = { bold: true };
}

function timelineSheet(workbook: ExcelJS.Workbook, project: ProjectDto) {
  const sheet = workbook.addWorksheet("Timeline");
  sheet.columns = [
    { header: "Phase", key: "name", width: 26 },
    { header: "Start", key: "start", width: 14 },
    { header: "End", key: "end", width: 14 },
    { header: "Duration (days)", key: "duration", width: 16 },
  ];
  sheet.getRow(1).font = { bold: true };
  for (const phase of project.phases) {
    const duration =
      phase.startDate && phase.endDate
        ? Math.round((new Date(phase.endDate).getTime() - new Date(phase.startDate).getTime()) / 86_400_000) + 1
        : "";
    sheet.addRow({
      name: phase.name,
      start: phase.startDate ? new Date(phase.startDate).toLocaleDateString() : "",
      end: phase.endDate ? new Date(phase.endDate).toLocaleDateString() : "",
      duration,
    });
  }
}

function storiesSheet(workbook: ExcelJS.Workbook, project: ProjectDto) {
  const sheet = workbook.addWorksheet("Stories");
  sheet.columns = [
    { header: "Epic", key: "epic", width: 22 },
    { header: "Story", key: "title", width: 40 },
    { header: "Description", key: "description", width: 60 },
    { header: "Acceptance Criteria", key: "ac", width: 60 },
    { header: "Story Points", key: "points", width: 12 },
    { header: "Phase", key: "phase", width: 22 },
  ];
  sheet.getRow(1).font = { bold: true };
  const phaseNameById = new Map(project.phases.map((p) => [p.id, p.name]));
  for (const story of project.stories) {
    sheet.addRow({
      epic: story.epic,
      title: story.title,
      description: story.description,
      ac: story.acceptanceCriteria.map((c) => `- ${c}`).join("\n"),
      points: story.storyPoints,
      phase: story.phaseId ? phaseNameById.get(story.phaseId) ?? "" : "",
    });
  }
}

function tasksSheet(workbook: ExcelJS.Workbook, project: ProjectDto) {
  const sheet = workbook.addWorksheet("Tasks");
  sheet.columns = [
    { header: "Epic", key: "epic", width: 22 },
    { header: "Story", key: "story", width: 40 },
    { header: "Task", key: "title", width: 40 },
    { header: "Hours", key: "hours", width: 10 },
    { header: "Role", key: "role", width: 20 },
  ];
  sheet.getRow(1).font = { bold: true };
  for (const story of project.stories) {
    for (const task of story.tasks) {
      sheet.addRow({ epic: story.epic, story: story.title, title: task.title, hours: task.hours, role: task.role });
    }
  }
}

function testCasesSheet(workbook: ExcelJS.Workbook, project: ProjectDto) {
  const sheet = workbook.addWorksheet("Test Cases");
  sheet.columns = [
    { header: "Test Case ID", key: "id", width: 14 },
    { header: "Story", key: "story", width: 40 },
    { header: "Title", key: "title", width: 40 },
    { header: "Precondition", key: "precondition", width: 30 },
    { header: "Steps", key: "steps", width: 50 },
    { header: "Expected Result", key: "expected", width: 40 },
    { header: "Priority", key: "priority", width: 10 },
    { header: "Type", key: "type", width: 14 },
  ];
  sheet.getRow(1).font = { bold: true };
  let counter = 0;
  for (const story of project.stories) {
    for (const tc of story.testCases) {
      counter += 1;
      sheet.addRow({
        id: `TC-${String(counter).padStart(3, "0")}`,
        story: story.title,
        title: tc.title,
        precondition: tc.precondition ?? "",
        steps: tc.steps.map((s, i) => `${i + 1}. ${s}`).join("\n"),
        expected: tc.expectedResult,
        priority: tc.priority,
        type: tc.type,
      });
    }
  }
}

export async function buildEstimateWorkbook(project: ProjectDto): Promise<ExcelJS.Buffer> {
  const workbook = new ExcelJS.Workbook();
  phaseBreakdownSheet(workbook, project);
  roleHoursSheet(workbook, project);
  timelineSheet(workbook, project);
  return workbook.xlsx.writeBuffer();
}

export async function buildStoriesWorkbook(project: ProjectDto): Promise<ExcelJS.Buffer> {
  const workbook = new ExcelJS.Workbook();
  storiesSheet(workbook, project);
  tasksSheet(workbook, project);
  return workbook.xlsx.writeBuffer();
}

export async function buildTestCasesWorkbook(project: ProjectDto): Promise<ExcelJS.Buffer> {
  const workbook = new ExcelJS.Workbook();
  testCasesSheet(workbook, project);
  return workbook.xlsx.writeBuffer();
}

export async function buildAllWorkbook(project: ProjectDto): Promise<ExcelJS.Buffer> {
  const workbook = new ExcelJS.Workbook();
  phaseBreakdownSheet(workbook, project);
  roleHoursSheet(workbook, project);
  timelineSheet(workbook, project);
  storiesSheet(workbook, project);
  tasksSheet(workbook, project);
  testCasesSheet(workbook, project);
  return workbook.xlsx.writeBuffer();
}

export function buildEstimateCsv(project: ProjectDto): string {
  const header = ["Phase", "Hours", "Buffer %", "Effective Hours", "Start", "End", "Rationale", "Dependencies"];
  const rows = project.phases.map((phase) => [
    phase.name,
    String(phase.hours),
    String(phase.bufferPercent),
    effectiveHours(phase.hours, phase.bufferPercent).toFixed(1),
    phase.startDate ? new Date(phase.startDate).toLocaleDateString() : "",
    phase.endDate ? new Date(phase.endDate).toLocaleDateString() : "",
    phase.rationale ?? "",
    phase.dependencies ?? "",
  ]);
  const escape = (v: string) => (/[",\n]/.test(v) ? `"${v.replace(/"/g, '""')}"` : v);
  return [header, ...rows].map((row) => row.map(escape).join(",")).join("\r\n");
}

export async function buildStoriesDocx(project: ProjectDto): Promise<Buffer> {
  const phaseNameById = new Map(project.phases.map((p) => [p.id, p.name]));
  const epics = new Map<string, typeof project.stories>();
  for (const story of project.stories) {
    if (!epics.has(story.epic)) epics.set(story.epic, []);
    epics.get(story.epic)!.push(story);
  }

  const children: (Paragraph | Table)[] = [
    new Paragraph({ text: project.name, heading: HeadingLevel.TITLE }),
    new Paragraph({ text: "Stories & Tasks", heading: HeadingLevel.HEADING_1 }),
  ];

  for (const [epicName, stories] of epics) {
    children.push(new Paragraph({ text: epicName, heading: HeadingLevel.HEADING_2 }));
    for (const story of stories) {
      children.push(
        new Paragraph({ text: `${story.title} (${story.storyPoints} pts)`, heading: HeadingLevel.HEADING_3 }),
        new Paragraph({ children: [new TextRun({ text: story.description, italics: true })] }),
        new Paragraph({ text: "Acceptance Criteria:", spacing: { before: 100 } }),
        ...story.acceptanceCriteria.map((c) => new Paragraph({ text: `• ${c}` })),
        new Paragraph({ text: `Phase: ${story.phaseId ? phaseNameById.get(story.phaseId) ?? "" : "Unassigned"}`, spacing: { before: 100 } })
      );

      if (story.tasks.length > 0) {
        children.push(
          new Table({
            width: { size: 100, type: WidthType.PERCENTAGE },
            rows: [
              new TableRow({
                children: ["Task", "Hours", "Role"].map(
                  (h) => new TableCell({ children: [new Paragraph({ children: [new TextRun({ text: h, bold: true })] })] })
                ),
              }),
              ...story.tasks.map(
                (task) =>
                  new TableRow({
                    children: [task.title, String(task.hours), task.role].map(
                      (v) => new TableCell({ children: [new Paragraph(v)] })
                    ),
                  })
              ),
            ],
          }),
          new Paragraph({ text: "" })
        );
      }
    }
  }

  const doc = new Document({ sections: [{ children }] });
  return Packer.toBuffer(doc);
}

File: streamlit_app/export_service.py
from __future__ import annotations

import csv
import io

from openpyxl import Workbook
from openpyxl.styles import Font
from docx import Document as DocxDocument

from constants import DEFAULT_ROLES
from models import ProjectDto


def _effective_hours(hours: float, buffer_percent: float) -> float:
    return hours * (1 + buffer_percent / 100)


def _fmt_date(iso: str | None) -> str:
    if not iso:
        return ""
    return iso[:10]


def _phase_breakdown_sheet(wb: Workbook, project: ProjectDto):
    sheet = wb.create_sheet("Phase Breakdown")
    headers = ["Phase", "Hours", "Buffer %", "Effective Hours", "Start", "End", "Rationale", "Dependencies"]
    sheet.append(headers)
    for cell in sheet[1]:
        cell.font = Font(bold=True)

    for phase in project.phases:
        sheet.append([
            phase.name,
            phase.hours,
            phase.bufferPercent,
            round(_effective_hours(phase.hours, phase.bufferPercent), 1),
            _fmt_date(phase.startDate),
            _fmt_date(phase.endDate),
            phase.rationale or "",
            phase.dependencies or "",
        ])

    total_row = ["Total", sum(p.hours for p in project.phases), "", round(sum(_effective_hours(p.hours, p.bufferPercent) for p in project.phases), 1), "", "", "", ""]
    sheet.append(total_row)
    for cell in sheet[sheet.max_row]:
        cell.font = Font(bold=True)

    widths = [26, 10, 10, 16, 14, 14, 50, 30]
    for i, w in enumerate(widths, start=1):
        sheet.column_dimensions[sheet.cell(row=1, column=i).column_letter].width = w


def _role_hours_sheet(wb: Workbook, project: ProjectDto):
    sheet = wb.create_sheet("Role Hours")
    roles = list(dict.fromkeys([*DEFAULT_ROLES, *project.customRoles, *[rh.role for p in project.phases for rh in p.roleHours]]))
    sheet.append(["Phase", *roles, "Total"])
    for cell in sheet[1]:
        cell.font = Font(bold=True)

    role_totals = {r: 0.0 for r in roles}
    for phase in project.phases:
        row = [phase.name]
        phase_total = 0.0
        for role in roles:
            hours = next((rh.hours for rh in phase.roleHours if rh.role == role), 0)
            row.append(hours)
            role_totals[role] += hours
            phase_total += hours
        row.append(phase_total)
        sheet.append(row)

    total_row = ["Total", *[role_totals[r] for r in roles], sum(role_totals.values())]
    sheet.append(total_row)
    for cell in sheet[sheet.max_row]:
        cell.font = Font(bold=True)

    sheet.column_dimensions["A"].width = 26


def _timeline_sheet(wb: Workbook, project: ProjectDto):
    sheet = wb.create_sheet("Timeline")
    sheet.append(["Phase", "Start", "End", "Duration (days)"])
    for cell in sheet[1]:
        cell.font = Font(bold=True)

    for phase in project.phases:
        duration = ""
        if phase.startDate and phase.endDate:
            from datetime import date as _date
            start = _date.fromisoformat(phase.startDate[:10])
            end = _date.fromisoformat(phase.endDate[:10])
            duration = (end - start).days + 1
        sheet.append([phase.name, _fmt_date(phase.startDate), _fmt_date(phase.endDate), duration])

    sheet.column_dimensions["A"].width = 26


def _stories_sheet(wb: Workbook, project: ProjectDto):
    sheet = wb.create_sheet("Stories")
    sheet.append(["Epic", "Story", "Description", "Acceptance Criteria", "Story Points", "Phase"])
    for cell in sheet[1]:
        cell.font = Font(bold=True)

    phase_name_by_id = {p.id: p.name for p in project.phases}
    for story in project.stories:
        sheet.append([
            story.epic,
            story.title,
            story.description,
            "\n".join(f"- {c}" for c in story.acceptanceCriteria),
            story.storyPoints,
            phase_name_by_id.get(story.phaseId, "") if story.phaseId else "",
        ])

    widths = [22, 40, 60, 60, 12, 22]
    for i, w in enumerate(widths, start=1):
        sheet.column_dimensions[sheet.cell(row=1, column=i).column_letter].width = w


def _tasks_sheet(wb: Workbook, project: ProjectDto):
    sheet = wb.create_sheet("Tasks")
    sheet.append(["Epic", "Story", "Task", "Hours", "Role"])
    for cell in sheet[1]:
        cell.font = Font(bold=True)

    for story in project.stories:
        for task in story.tasks:
            sheet.append([story.epic, story.title, task.title, task.hours, task.role])

    widths = [22, 40, 40, 10, 20]
    for i, w in enumerate(widths, start=1):
        sheet.column_dimensions[sheet.cell(row=1, column=i).column_letter].width = w


def _test_cases_sheet(wb: Workbook, project: ProjectDto):
    sheet = wb.create_sheet("Test Cases")
    sheet.append(["Test Case ID", "Story", "Title", "Precondition", "Steps", "Expected Result", "Priority", "Type"])
    for cell in sheet[1]:
        cell.font = Font(bold=True)

    counter = 0
    for story in project.stories:
        for tc in story.testCases:
            counter += 1
            sheet.append([
                f"TC-{counter:03d}",
                story.title,
                tc.title,
                tc.precondition or "",
                "\n".join(f"{i + 1}. {s}" for i, s in enumerate(tc.steps)),
                tc.expectedResult,
                tc.priority,
                tc.type,
            ])

    widths = [14, 40, 40, 30, 50, 40, 10, 14]
    for i, w in enumerate(widths, start=1):
        sheet.column_dimensions[sheet.cell(row=1, column=i).column_letter].width = w


def _workbook_bytes(wb: Workbook) -> bytes:
    wb.remove(wb["Sheet"])
    buf = io.BytesIO()
    wb.save(buf)
    return buf.getvalue()


def build_estimate_workbook(project: ProjectDto) -> bytes:
    wb = Workbook()
    _phase_breakdown_sheet(wb, project)
    _role_hours_sheet(wb, project)
    _timeline_sheet(wb, project)
    return _workbook_bytes(wb)


def build_stories_workbook(project: ProjectDto) -> bytes:
    wb = Workbook()
    _stories_sheet(wb, project)
    _tasks_sheet(wb, project)
    return _workbook_bytes(wb)


def build_test_cases_workbook(project: ProjectDto) -> bytes:
    wb = Workbook()
    _test_cases_sheet(wb, project)
    return _workbook_bytes(wb)


def build_all_workbook(project: ProjectDto) -> bytes:
    wb = Workbook()
    _phase_breakdown_sheet(wb, project)
    _role_hours_sheet(wb, project)
    _timeline_sheet(wb, project)
    _stories_sheet(wb, project)
    _tasks_sheet(wb, project)
    _test_cases_sheet(wb, project)
    return _workbook_bytes(wb)


def build_estimate_csv(project: ProjectDto) -> str:
    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow(["Phase", "Hours", "Buffer %", "Effective Hours", "Start", "End", "Rationale", "Dependencies"])
    for phase in project.phases:
        writer.writerow([
            phase.name,
            phase.hours,
            phase.bufferPercent,
            round(_effective_hours(phase.hours, phase.bufferPercent), 1),
            _fmt_date(phase.startDate),
            _fmt_date(phase.endDate),
            phase.rationale or "",
            phase.dependencies or "",
        ])
    return buf.getvalue()


def build_stories_docx(project: ProjectDto) -> bytes:
    phase_name_by_id = {p.id: p.name for p in project.phases}
    epics: dict[str, list] = {}
    for story in project.stories:
        epics.setdefault(story.epic, []).append(story)

    doc = DocxDocument()
    doc.add_heading(project.name, level=0)
    doc.add_heading("Stories & Tasks", level=1)

    for epic_name, stories in epics.items():
        doc.add_heading(epic_name, level=2)
        for story in stories:
            doc.add_heading(f"{story.title} ({story.storyPoints} pts)", level=3)
            desc = doc.add_paragraph()
            desc.add_run(story.description).italic = True

            doc.add_paragraph("Acceptance Criteria:")
            for c in story.acceptanceCriteria:
                doc.add_paragraph(c, style="List Bullet")

            phase_name = phase_name_by_id.get(story.phaseId, "Unassigned") if story.phaseId else "Unassigned"
            doc.add_paragraph(f"Phase: {phase_name}")

            if story.tasks:
                table = doc.add_table(rows=1, cols=3)
                table.style = "Light Grid Accent 1"
                hdr = table.rows[0].cells
                hdr[0].text, hdr[1].text, hdr[2].text = "Task", "Hours", "Role"
                for task in story.tasks:
                    row = table.add_row().cells
                    row[0].text, row[1].text, row[2].text = task.title, str(task.hours), task.role
            doc.add_paragraph("")

    buf = io.BytesIO()
    doc.save(buf)
    return buf.getvalue()
